### ISSUE TITLE
R4.2 , Rstan & Windows  

### ISSUE THREAD

**Issue by rok-cesnovar** [stan-dev/rstan#1006]
I am trying to make a blog post on what Windows users should do with R4.2, which is accompanied by a new build toolchain Rtools42. 

I am a bit lost on how to get rstan working, so I can include some instructions for rstan - the general recommendation will be to hold off on upgrading, if possible - as we wait for the first wave of issues to pass.

I set up a clean Windows instance, installed R4.2 and then just installing rstan (both the CRAN as well as the r-packages) version.
There is no .Renviron or .Rprofile or whatever - this is a clean install.

Rstan installs fine, but I am unable to build any model. I get the following error:

```
Error in compileCode(f, code, language = language, verbose = verbose) : 
  C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: file16cc56ad665a.o:file16cc56ad665a.cpp:(.text$_ZN3tbb8internal26task_scheduler_observer_v3D0Ev[_ZN3tbb8internal26task_scheduler_observer_v3D0Ev]+0x1d): undefined reference to `tbb::internal::task_scheduler_observer_v3::observe(bool)'C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: file16cc56ad665a.o:file16cc56ad665a.cpp:(.text$_ZN3tbb10interface623task_scheduler_observerD1Ev[_ZN3tbb10interface623task_scheduler_observerD1Ev]+0x1d): undefined reference to `tbb::internal::task_scheduler_observer_v3::observe(bool)'C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: file16cc56ad665a.o:file16cc56ad665a.cpp:(.text$_ZN3tbb10interface623task_scheduler_observerD1Ev[_ZN3tbb10interface623task_scheduler_observerD1Ev]+0x3a): undefined reference to `tbb::internal::task_scheduler_observer_v3::observe(bool)'C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: file16cc56ad665a.o:file16cc56ad665a.cpp:(.text$_ZN3tbb10interface
Error in sink(type = "output") : invalid connection
```

Rcpparallel flags are:
```
> RcppParallel::LdFlags()
-LC:/PROGRA~1/R/R-42~1.0RC/library/RCPPPA~1/lib/x64 -ltbb -ltbbmalloc> 
```

Does anyone have any ideas on how to debug next?

Tagging @hsbadr @andrjohns, you guys typically know what to do in these cases :)

**Comment by hsbadr** [comment]
```C++
undefined reference to `tbb::internal::task_scheduler_observer_v3::observe(bool)
```
It seems that RStan plugin flags (including `RcppParallel` LDFLAGS) aren't used when `inline::compileCode()` is invoked. So, compilation is failed because it doesn't link to TBB, which is now required by Stan/Math.

**Comment by rok-cesnovar** [comment]
Thanks! Can I do anything to address this or does this require a fix in rstan?

**Comment by hsbadr** [comment]
Not sure if this is an RStan issue or something within `pkgbuild::with_build_tools()` or `inline::cxxfunction()`. I'll look into it.

**Comment by hsbadr** [comment]
Here's the backtrace:
```r
Backtrace:
    ▆
 1. └─rstan::stan_model("schools.stan", verbose = TRUE)
 2.   └─rstan:::cxxfunctionplus(...) at rstan/R/rstan.R:177:2
 3.     ├─pkgbuild::with_build_tools(...) at rstan/R/cxxfunplus.R:174:2
 4.     │ └─withr::with_path(rtools_path(), code) at pkgbuild/R/build-tools.R:66:4
 5.     │   └─base::force(code)
 6.     └─inline::cxxfunction(...) at rstan/R/cxxfunplus.R:174:2
 7.       └─inline:::compileCode(f, code, language = language, verbose = verbose) at inline/R/cxxfunction.R:130:4
 ```
RStan plugin environment settings are set correctly:
```r
>> setting environment variables: 
LOCAL_LIBS =  "C:/Users/hsbadr/Documents/R/library/rstan/lib/x64/libStanServices.a" -L"C:/Users/hsbadr/Documents/R/library/StanHeaders/libs/x64" -lStanHeaders -L"C:/Users/hsbadr/Documents/R/library/RcppParallel/lib/x64" -ltbb -LC:/Users/hsbadr/Documents/R/library/RcppParallel/lib/x64 -ltbb -ltbbmalloc
PKG_CPPFLAGS =   -I"C:/Users/hsbadr/Documents/R/library/Rcpp/include/"  -I"C:/Users/hsbadr/Documents/R/library/RcppEigen/include/"  -I"C:/Users/hsbadr/Documents/R/library/RcppEigen/include/unsupported"  -I"C:/Users/hsbadr/Documents/R/library/BH/include" -I"C:/Users/hsbadr/Documents/R/library/StanHeaders/include/src/"  -I"C:/Users/hsbadr/Documents/R/library/StanHeaders/include/"  -I"C:/Users/hsbadr/Documents/R/library/RcppParallel/include/" -DRCPP_PARALLEL_USE_TBB=1 -I"C:/Users/hsbadr/Documents/R/library/rstan/include" -DEIGEN_NO_DEBUG  -DBOOST_DISABLE_ASSERTS  -DBOOST_PENDING_INTEGER_LOG2_HPP  -DSTAN_THREADS  -DUSE_STANC3 -DSTRICT_R_HEADERS  -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION  -DBOOST_NO_AUTO_PTR  -include "C:/Users/hsbadr/Documents/R/library/StanHeaders/include/stan/math/prim/fun/Eigen.hpp"  -std=c++1y
```
but not used:
```r
g++  -shared -s -static-libgcc -o file300853c657ab.dll tmp.def file300853c657ab.o  -L"C:/rtools42/x86_64-w64-mingw32.static.posix/lib/x64" -L"C:/rtools42/x86_64-w64-mingw32.static.posix/lib"  -L"C:/PROGRA~1/R/R-devel/bin/x64" -lR
```

So, it's likely that `make` ignores the environment flags on Windows for some reason.

**Comment by hsbadr** [comment]
Ok. I think I can fix this in RStan.

**Comment by hsbadr** [comment]
@rok-cesnovar Should be fixed in [RStan v2.26.11](https://github.com/stan-dev/r-packages/commit/03214e4f72c809314ac17f96988f64004a86560b).

**Comment by rok-cesnovar** [comment]
Thanks a ton! All set now.

For anyone facing the issue in the top post:

The CRAN version of rstan currently does not work with R 4.2 on Windows. You need to install the Stan-hosted version of the package, which can be installed by running the following:

```r
remove.packages(c("rstan", "StanHeaders"))
install.packages("rstan", repos = c("https://mc-stan.org/r-packages/", getOption("repos")))
```
